doc/HttpTransportAdapter.md

```diff
@@ -0,0 +1,223 @@
+
+Follow these steps to implement your own HTTP transport adapter:
+
+1. Implement the interface.
+
+The HTTP transport adapter interface is define in `Azure::Core::HTTP::HttpTransport` from `inc/azure/core/http/transport.hpp`.
+
+```cpp
+/*
+* Implementing the HTTP transport adapter.
+*/
+
+#include "azure/core/http/transport.hpp"
+
+// Derive from HttpTransport
+class CustomTransportAdapter : public Azure::Core::Http::HttpTransport {
+  public:
+    // Override the Send method
+    std::unique_ptr<Azure::Core::Http::RawResponse> Send(
+        Azure::Core::Context const& context,
+        Azure::Core::Http::Request& request) override;
+};
+```
+
+> The implementation needs to take care of using the members from the `Request` argument which contains all the HTTP request information to write the HTTP request using some C++ HTTP stack. Then it needs to create the `RawResponse` from the response and return it. The `context` argument can be optionally used to check if the request is cancelled before performing I/O operations (like writing to the network).
+
+2. Stateless component.
+
+Make sure that the HTTP transport adapter that is created is stateless. This is because it can be shared from multiple SDK clients as a shared ptr. Consider designing a session object, for instance, if you need to hold some state. See the next pattern as an example.
+
+```cpp
+/*
+* Using a session pattern for the HTTP transport adapter.
+*/
+
+// Create a session class first.
+class CustomHttpTransportAdapterSession {
+  private:
+    // Keep any state like network/socket handlers here.
+    SomeHttpClientLibraryHandler handler;
+    SomeOtherRequiredState state;
+  
+  public:
+    // Expose any functionality required. For example, this could
+    // be the main method to produce the HTTP raw response.
+    std::unique_ptr<Azure::Core::Http::RawResponse> Perform(
+      Azure::Core::Context const& context,
+      Azure::Core::Http::Request& request
+    );
+}
+```
+
+Then, within the Send() method implementation.
+
+```cpp
+// Create a new Session object when calling Send().
+    std::unique_ptr<Azure::Core::Http::RawResponse> Send(
+        Azure::Core::Context const& context,
+        Azure::Core::Http::Request& request) {
+
+          // This will ensure that multiple clients can
+          // call Send() at the same time.
+          CustomHttpTransportAdapterSession session;
+
+          // Use any functionally from your created session.
+          // For this example, we could just return like this:
+          return session.Perform();
+        }
+
+```
+
+> If instead of having a session created for every call to Send(), we would place the state directly into the HTTP transport adapter class, any new call to Send() would corrupt the current state when sharing the same HTTP transport adapter from multiple clients.
```

Comment:
I believe that isn't necessarily true, especially with WinHTTP, once we address this: https://github.com/Azure/azure-sdk-for-cpp/issues/1098

Maybe remove this for now?

Reply:
#1098 is a different issue.

This note if to mention what happen if you don't design the adapter in a multi-thread shared way, where it would be corrupted from a second call

Reply:
I think overloading the term "session" here is what's causing the confusion. In WinHTTP, I may not create a **session** (i.e. session handle) for every call to `Send()`.The way it is written, it makes it seem like if I don't create a new session on each call to Send(), I am corrupting some state. How is that always true?

Reply:
Correct. Session happens to be the same name as what WinHTTP main component uses.

For WinHTTP, the equivalent would be the `handleManager`. You create a new handleManager  per each Send() call and init the handlers inside it.
If you would have the handlers directly into the transport adapter class, a second call to Send() would interfere collapse if there is currently one Send() happening because it would try to use the same handlers.

Re-using a connection handler is possible, yes. That's what #1098 is all about. However, the connection can't be re-used while it is on use. The same handler can be use again only when the Send() method is finished and there's a new Send() call that happen to have the same configuration used before from a previous call to Send().